### PR TITLE
An All-In-One Spring Security Crash Course For Java Developers

### LINKED ISSUE #1244: Spring Security Crash Course
### Brief Summary: 
A crash course of some of the major features Spring security provides to secure your Spring boot applications. This guide will go over a few authentication strategies, ways to add authorization and more.

### Key Takeaways: 
- What is authentication and authorization
- The theory behind http basic, JWT, and OAuth authentication and how to implement them
- How to implement role based authorization
- Method level authorization with the preauthorize and postauthorize annotations

### References: 
N/A

### PR DESCRIPTION
**Pull request by john-amiscaray**
This fixes #1244
### Article Checklist
**NOTE: (1st-time contributors - submit ONLY one article to get familiar with the process, quality, and style-once published, you can then submit any number of articles.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by paulodhiambo** [comment]
@hectorkambow I wish to review this article.
Regards.

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
@john-amiscaray amazing tutorial on Spring security. I have mentioned a few suggestions, review those whenever you get a chance and ping me when done.

File: articles/an-all-in-one-spring-security-crash-course-for-java-developers/index.md [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+
+Whenever we build an application, security isn’t exactly the first thing we think about. Yet, it is one of the most important features that needs your attention. In this article, we will cover some key ways to secure your apps with Spring Security. I know security might seem a little daunting especially if you are focused on the bigger picture. Knowing this I tried to make a simple all-in-one guide that would have helped me when I was learning this.
+
+### Prerequisites
+
+In this guide, I will assume that you have experience with Spring Boot. This includes some of the common design patterns such as DTOs and services. As well, I will assume you have some basic knowledge of Spring data to store users in a database.
+
+### Theory
+
+Before we begin coding, let’s go over some theory. The main features we will cover are different tactics for authentication and authorization. Authorization is the process of verifying we allow the user to do what they are asking to do. Authentication is the process of verifying who it is that is sending a request. The three authentication strategies we will go over are HTTP basic, JWT, and OAuth.
```
[the remark this answers is not in the document]
Comment (john-amiscaray) [reply articles/an-all-in-one-spring-security-crash-course-for-java-developers/index.md:16]: The reason I referred to it as theory and not Spring security fundamentals is that that content is more so about general web security concepts that can be applied outside of Spring so I don't believe I want to make that change.

File: articles/an-all-in-one-spring-security-crash-course-for-java-developers/index.md [hunk 2]
```diff
@@ -0,0 +1,518 @@
+Title: An All-In-One Spring Security Crash Course For Java Developers
+
+Description: A crash course on how to secure a Spring Boot application. 
+
+
+### Introduction
+
+Whenever we build an application, security isn’t exactly the first thing we think about. Yet, it is one of the most important features that needs your attention. In this article, we will cover some key ways to secure your apps with Spring Security. I know security might seem a little daunting especially if you are focused on the bigger picture. Knowing this I tried to make a simple all-in-one guide that would have helped me when I was learning this.
```
[the remark this answers is not in the document]
Comment (john-amiscaray) [reply articles/an-all-in-one-spring-security-crash-course-for-java-developers/index.md]: @paulodhiambo Okay I have just committed that change. Thank you.

**Review by paulodhiambo — APPROVED** [review]

**Comment by paulodhiambo** [comment]
@john-amiscaray I am proceeding to tech review.

**Comment by paulodhiambo** [comment]
@john-amiscaray amazing article on Spring security.

**Comment by john-amiscaray** [comment]
> 
> 
> @john-amiscaray amazing article on Spring security.

Thank you for taking the time to review this.

**Comment by hectorkambow** [comment]
plagiarism check done

**Pull request merged**